### PR TITLE
NIFI-3717: Added HortonworksSchemaRegistry

### PR DESCRIPTION
**Pull request by markap14**
Thank you for submitting a contribution to Apache NiFi.

In order to streamline the review of the contribution we ask you
to ensure the following steps have been taken:

### For all changes:
- [ ] Is there a JIRA ticket associated with this PR? Is it referenced 
     in the commit message?

- [ ] Does your PR title start with NIFI-XXXX where XXXX is the JIRA number you are trying to resolve? Pay particular attention to the hyphen "-" character.

- [ ] Has your PR been rebased against the latest commit within the target branch (typically master)?

- [ ] Is your initial contribution a single, squashed commit?

### For code changes:
- [ ] Have you ensured that the full suite of tests is executed via mvn -Pcontrib-check clean install at the root nifi folder?
- [ ] Have you written or updated unit tests to verify your changes?
- [ ] If adding new dependencies to the code, are these dependencies licensed in a way that is compatible for inclusion under [ASF 2.0](http://www.apache.org/legal/resolved.html#category-a)? 
- [ ] If applicable, have you updated the LICENSE file, including the main LICENSE file under nifi-assembly?
- [ ] If applicable, have you updated the NOTICE file, including the main NOTICE file found under nifi-assembly?
- [ ] If adding new Properties, have you added .displayName in addition to .name (programmatic access) for each of the new properties?

### For documentation related changes:
- [ ] Have you ensured that format looks appropriate for the output in which it is rendered?

### Note:
Please ensure that once the PR is submitted, you check travis-ci for build issues and submit an update to your PR as soon as possible.

### CODE REVIEW

File: nifi-nar-bundles/nifi-standard-services/nifi-hwx-schema-registry-bundle/nifi-hwx-schema-registry-nar/pom.xml [hunk 1]
```diff
@@ -0,0 +1,39 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- 
+    Copyright 2016 Hortoworks, Inc. All rights reserved.
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-standard-services/nifi-hwx-schema-registry-bundle/nifi-hwx-schema-registry-nar/pom.xml]: this is not correct.  This is apache nifi code.  This header would only make sense if we're copying source from hortonworks repos.

File: nifi-nar-bundles/nifi-standard-services/nifi-hwx-schema-registry-bundle/nifi-hwx-schema-registry-service/pom.xml [hunk 2]
```diff
@@ -0,0 +1,195 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!-- 
+    Copyright 2016 Hortoworks, Inc. All rights reserved.
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-standard-services/nifi-hwx-schema-registry-bundle/nifi-hwx-schema-registry-service/pom.xml]: must fix this to be the proper apache nifi header.

File: nifi-nar-bundles/nifi-standard-services/nifi-hwx-schema-registry-bundle/nifi-hwx-schema-registry-service/src/main/java/org/apache/nifi/schemaregistry/hortonworks/HortonworksSchemaRegistry.java [hunk 3]
```diff
@@ -0,0 +1,323 @@
...
+import com.hortonworks.registries.schemaregistry.SchemaMetadata;
+import com.hortonworks.registries.schemaregistry.SchemaMetadataInfo;
+import com.hortonworks.registries.schemaregistry.SchemaVersionInfo;
+import com.hortonworks.registries.schemaregistry.SchemaVersionKey;
+import com.hortonworks.registries.schemaregistry.client.SchemaRegistryClient;
+import com.hortonworks.registries.schemaregistry.errors.SchemaNotFoundException;
+
+@Tags({"schema", "registry", "avro", "hortonworks", "hwx"})
+@CapabilityDescription("Provides a Schema Registry Service that interacts with a Hortonworks Schema Registry")
+public class HortonworksSchemaRegistry extends AbstractControllerService implements SchemaRegistry {
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-standard-services/nifi-hwx-schema-registry-bundle/nifi-hwx-schema-registry-service/src/main/java/org/apache/nifi/schemaregistry/hortonworks/HortonworksSchemaRegistry.java]: Do you think perhaps this makes more sense as 'HortonworksSchemaRegistryClient' ?    The name without it sort of implies this IS the hortonworks schema registry.

File: nifi-nar-bundles/nifi-standard-services/nifi-hwx-schema-registry-bundle/nifi-hwx-schema-registry-service/src/main/java/org/apache/nifi/schemaregistry/hortonworks/HortonworksSchemaRegistry.java [hunk 4]
```diff
@@ -0,0 +1,323 @@
...
+
+import com.hortonworks.registries.schemaregistry.SchemaMetadata;
+import com.hortonworks.registries.schemaregistry.SchemaMetadataInfo;
+import com.hortonworks.registries.schemaregistry.SchemaVersionInfo;
+import com.hortonworks.registries.schemaregistry.SchemaVersionKey;
+import com.hortonworks.registries.schemaregistry.client.SchemaRegistryClient;
+import com.hortonworks.registries.schemaregistry.errors.SchemaNotFoundException;
+
+@Tags({"schema", "registry", "avro", "hortonworks", "hwx"})
+@CapabilityDescription("Provides a Schema Registry Service that interacts with a Hortonworks Schema Registry")
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-standard-services/nifi-hwx-schema-registry-bundle/nifi-hwx-schema-registry-service/src/main/java/org/apache/nifi/schemaregistry/hortonworks/HortonworksSchemaRegistry.java]: i think we should link to the github page for the hwx schema registry project here.

File: nifi-nar-bundles/nifi-standard-services/nifi-hwx-schema-registry-bundle/nifi-hwx-schema-registry-service/src/main/resources/META-INF/services/org.apache.nifi.controller.ControllerService [hunk 5]
```diff
@@ -0,0 +1,12 @@
+#  Copyright 2016 Hortoworks, Inc. All rights reserved.
```
Comment (joewitt) [inline nifi-nar-bundles/nifi-standard-services/nifi-hwx-schema-registry-bundle/nifi-hwx-schema-registry-service/src/main/resources/META-INF/services/org.apache.nifi.controller.ControllerService]: this is not the correct source header.

**Comment by joewitt** [comment]
[INFO] --- maven-checkstyle-plugin:2.15:check (check-style) @ nifi-hwx-schema-registry-service ---
[WARNING] src/main/java/org/apache/nifi/schemaregistry/hortonworks/HortonworksSchemaRegistry.java[103:9] (whitespace) FileTabCharacter: Line contains a tab character.
[WARNING] src/main/java/org/apache/nifi/schemaregistry/hortonworks/HortonworksSchemaRegistry.java[217:8] (javadoc) JavadocMethod: Unused @param tag for 'schemaName'.
[INFO] ------------------------------------------------------------------------

File: nifi-nar-bundles/nifi-standard-services/nifi-hwx-schema-registry-bundle/nifi-hwx-schema-registry-nar/pom.xml (hunk 1 above)
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-services/nifi-hwx-schema-registry-bundle/nifi-hwx-schema-registry-nar/pom.xml]: Sorry - you're right. I fixed this in all my java code but forgot that this stuff automatically gets added to XML files and the like as well.

File: nifi-nar-bundles/nifi-standard-services/nifi-hwx-schema-registry-bundle/nifi-hwx-schema-registry-service/src/main/java/org/apache/nifi/schemaregistry/hortonworks/HortonworksSchemaRegistry.java (hunk 3 above)
Comment (markap14) [inline nifi-nar-bundles/nifi-standard-services/nifi-hwx-schema-registry-bundle/nifi-hwx-schema-registry-service/src/main/java/org/apache/nifi/schemaregistry/hortonworks/HortonworksSchemaRegistry.java]: I could see either way. As a user, I think of myself referencing the service as if it were the Hortonworks Schema Registry itself.... we could change it if you feel it's necessary.

**Comment by markap14** [comment]
@joewitt I think I've addressed all of the comments above, except for the naming. Let me know if you think we need to change the class name.

**Comment by joewitt** [comment]
i'm fine with the name and see your point.  I've updated the L&N a bit but honestly that was really darn close so thanks!  I've done the full clean/build contrib check and tested against a live schema registry instance.  Very cool.  I was able to stream provenance events through and convert them between formats with minimal thought and at high rates.  That is awesome. +1 will merge to master.

**Pull request closed without merging**